core/lib/scmi-server/scmi_server_scpfw.c

```diff
@@ -0,0 +1,353 @@
+
+	ctx = calloc(1, sizeof(*ctx));
+	if (!ctx)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	ctx->dt_name = fdt_get_name(fdt, parent_node, NULL);
+
+	/* Read device tree */
+	SIMPLEQ_INIT(&ctx->agent_list);
+
+	fdt_for_each_subnode(agent_node, fdt, parent_node) {
+		const char *node_name = fdt_get_name(fdt, agent_node, NULL);
+
+		if (!strstr(node_name, "agent@"))
+			continue;
+
+		agent_ctx = calloc(1, sizeof(*agent_ctx));
+		if (!agent_ctx) {
+			res = TEE_ERROR_OUT_OF_MEMORY;
+			goto fail_agent;
+		}
+		agent_ctx->dt_name = node_name;
+
+		res = optee_scmi_server_probe_agent(fdt, agent_node, agent_ctx);
+		if (res)
+			goto fail_agent;
+
+		SIMPLEQ_FOREACH(a, &ctx->agent_list, link)
+			assert(a->agent_id != agent_ctx->agent_id);
+
+		SIMPLEQ_INSERT_TAIL(&ctx->agent_list, agent_ctx, link);
+		agent_cfg_count = MAX(agent_cfg_count, agent_ctx->agent_id);
+	}
+
+	agent_cfg_count++;
+
+	/* Create SCMI config structures */
+	scpfw_cfg.agent_count = agent_cfg_count;
+	scpfw_cfg.agent_config = calloc(scpfw_cfg.agent_count,
+					sizeof(*scpfw_cfg.agent_config));
+	if (!scpfw_cfg.agent_config) {
+		res = TEE_ERROR_OUT_OF_MEMORY;
+		scpfw_cfg.agent_count = 0;
+		goto fail_agent;
+	}
+
+	SIMPLEQ_FOREACH(agent_ctx, &ctx->agent_list, link) {
+		struct scpfw_agent_config *agent_cfg =
+			scpfw_cfg.agent_config + agent_ctx->agent_id;
+
+		agent_cfg->name = (const char *)strdup(agent_ctx->dt_name);
+		if (!agent_cfg->name) {
+			res = TEE_ERROR_OUT_OF_MEMORY;
+			goto fail_scpfw_cfg;
+		}
+
+		agent_cfg->agent_id = agent_ctx->agent_id;
+
+		/*
+		 * Right now this driver can handle one channel per agent only.
+		 */
+		assert(agent_ctx->channel_id == 0);
+		agent_cfg->channel_count = 1;
+		agent_cfg->channel_config =
+			calloc(agent_cfg->channel_count,
+			       sizeof(*agent_cfg->channel_config));
+		if (!agent_cfg->channel_config) {
+			res = TEE_ERROR_OUT_OF_MEMORY;
+			agent_cfg->channel_count = 0;
+			goto fail_scpfw_cfg;
+		}
+
+		for (i = 0; i < agent_cfg->channel_count; i++) {
+			struct scpfw_channel_config *channel_cfg =
+				agent_cfg->channel_config + i;
+
+			channel_cfg->name = "channel";
+			channel_cfg->channel_id = agent_ctx->channel_id;
+		}
+	}
+
+	/* Parse protocols and fill channels config */
+	SIMPLEQ_FOREACH(agent_ctx, &ctx->agent_list, link) {
+		struct optee_scmi_server_protocol *protocol_ctx = NULL;
+		struct scpfw_agent_config *agent_cfg =
+			scpfw_cfg.agent_config + agent_ctx->agent_id;
+		struct scpfw_channel_config *channel_cfg =
+			agent_cfg->channel_config + agent_ctx->channel_id;
+
+		SIMPLEQ_FOREACH(protocol_ctx, &agent_ctx->protocol_list, link) {
+			switch (protocol_ctx->protocol_id) {
+			case SCMI_PROTOCOL_ID_CLOCK:
+				res = optee_scmi_server_init_clocks(fdt,
+					protocol_ctx->dt_node, agent_cfg,
+					channel_cfg);
+				if (res)
+					panic("Error during clocks init");
+				break;
+			case SCMI_PROTOCOL_ID_RESET_DOMAIN:
+				res = optee_scmi_server_init_resets(fdt,
+					protocol_ctx->dt_node, agent_cfg,
```

Comment:
checkpatch complains about the alignment here and indeed this is indented quite a lot. Could the function be made shorter/simpler by using helper functions?